examples/client.rs

```diff
@@ -1,41 +1,100 @@
+#[structopt(name = "client")]
+pub struct Configuration {
+    /// Verbose mode
+    #[structopt(short, long)]
+    verbose: bool,
+
+    /// Input audio file
+    #[structopt(short, long, required=true)]
+    input: String,
+
+    /// Output audio file
+    #[structopt(short, long, required=true)]
+    output: String,
+}
+
 fn main() {
-    let audio_file_path = args().nth(1)
-        .expect("Please specify an audio file to run RNNoise on");
-    let audio_file = File::open(audio_file_path).unwrap();
+    let configuration = Configuration::from_args();
+    if configuration.verbose {
+        println!("Configuration: {:?}", configuration);
+    }
+
+    let audio_file = File::open(&configuration.input).unwrap();
     let mut reader = Reader::new(audio_file).unwrap();
     let desc = reader.description();
     assert_eq!(1, desc.channel_count(),
         "The channel count is required to be one, at least for now");
 
     // Obtain the buffer of samples
-    let audio_buf :Vec<_> = if desc.sample_rate() == SAMPLE_RATE {
-        reader.samples().map(|s| s.unwrap()).collect()
+    let mut audio_buf :Vec<_> = if desc.sample_rate() == SAMPLE_RATE {
+        reader.samples::<f32>().map(|s| s.unwrap()).collect()
     } else {
         // We need to interpolate to the target sample rate
-        let interpolator = Linear::new([0i16], [0]);
+        let interpolator = Linear::new([0f32], [0.0]);
         let conv = Converter::from_hz_to_hz(
-            from_iter(reader.samples::<i16>().map(|s| [s.unwrap()])),
+            from_iter(reader.samples::<f32>().map(|s| [s.unwrap()])),
             interpolator,
             desc.sample_rate() as f64,
             SAMPLE_RATE as f64);
         conv.until_exhausted().map(|v| v[0]).collect()
     };
 
-    let rnnoise = DenoiseState::new();
+    if configuration.verbose {
+        println!("{} length: {}", &configuration.input, audio_buf.len());
+    }
+
+    // The library requires each frame be exactly FRAME_SIZE, so we append
+    // some zeros to be sure the final frame is sufficiently long.
+    let padding = audio_buf.len() % FRAME_SIZE;
+    if padding > 0 {
+        let mut pad: Vec<f32> = vec![0.0; FRAME_SIZE - padding];
+        audio_buf.append(&mut pad);
+        if configuration.verbose {
+            println!("padded audio file with {} characters", padding);
+        }
+    }
+    let buffers = audio_buf[..].chunks(FRAME_SIZE);
+    let mut denoised_buffer: Vec<f32> = vec![];
+    let mut rnnoise = DenoiseState::new();
+    for buffer in buffers {
+        let mut denoised: Vec<f32> = vec![0.0; FRAME_SIZE];
+        rnnoise.process_frame_mut(&buffer, &mut denoised[..]);
+        denoised_buffer.append(&mut denoised);
+    }
+
+    if configuration.verbose {
+        println!("{} length: {}", &configuration.output, denoised_buffer.len());
+    }
+    assert_eq!(audio_buf.len(), denoised_buffer.len());
+
+    // Write denoised buffer into output file -- currently written as raw
```

Comment:
Can you use sth like claxon for saving in a proper format?

Reply:
claxon seems to be a decoder only? I'd very much like to write it in a proper format but have not yet figured out how. Any pointers would be appreciated.

Reply:
Oh sorry I've meant hound. It can encode to wav. Not sure if it supports the sampling rate needed or not.

Reply:
I tried with hound for a while and finally gave up -- I couldn't figure out how to get it to accept the raw audio as input

Reply:
one thought which I didn't try was writing it out in a temporary file and then re-reading it -- seemed a bit silly, but it seems to expect to load things from a file. That said, I likely will hit the same problems of it missing metadata.

Reply:
The hound docs have examples how to create and use a WavWriter: https://docs.rs/hound/3.4.0/hound/#examples

You need to specify the given sample rate and the number of channels and then you call `writer.write_sample` for each sample in the denoised buffer. It should be pretty straight forward.

Reply:
Where I get lost is how to implement the Sample trait. Setting up the writers is clearly demonstrated in the examples, but then they generate their own audio rather than loading a raw buffer. This is where my newness with Rust is going to shine through ...

Attempts to `writer.write_sample` with the raw audio results in errors like:
```
error: the trait bound `std::vec::Vec<f32>: hound::Sample` is not satisfied
label: the trait `hound::Sample` is not implemented for `std::vec::Vec<f32>`
```
https://docs.rs/hound/3.4.0/hound/trait.Sample.html

Reply:
I dug through users of hound until I found an example doing what I needed, and you're correct, it is simple:
https://github.com/Prior99/rust-samplerate/blob/fe62a815f3d600d0f439f6b60d4a2a81b6178e0c/examples/simple.rs#L22

An example running with the latest code:
```
 cargo run --example client -- -v -i input.wav -o output.wav
    Finished dev [unoptimized + debuginfo] target(s) in 0.05s
     Running `target/debug/examples/client -v -i input.wav -o output.wav`
Configuration: Configuration { verbose: true, input: "input.wav", output: "output.wav" }
input.wav length: 51273
padded audio file with 393 characters
output.wav length: 51360
```

And comparing the input and output files:
```
$ file *wav
input.wav:  RIFF (little-endian) data, WAVE audio, Microsoft PCM, 16 bit, mono 16000 Hz
output.wav: RIFF (little-endian) data, WAVE audio, mono 48000 Hz
```

Reply:
Perhaps an ideal example would return the wav file to the original settings (16-bit, mono 16_000 Hz)?